scanner/README.md

```diff
@@ -0,0 +1,69 @@
+# Scanner
+
+## Workflow (for v5, not current)
+
+In order of action:
+
+ - Scanner gets `/videos` & scan file system to list all new videos
+ - Scanner guesses as much as possible from filename/path ALONE (no external database query).
+   - Format should be:
+     ```json5
+     {
+         path: string,
+         version: number,
+         part: number | null,
+         rendering: sha(path except version & part),
+         guess: {
+             kind: movie | episode | trailer | interview | ...,
+             name: string,
+             year: number | null,
+             season?: number,
+             episode?: number,
+             ...
+          },
+     }
+     ```
+
+     - Apply remaps from lists (AnimeList + thexem). Format is now:
+     ```json5
+     {
+         path: string,
+         version: number,
+         part: number | null,
+         rendering: sha(path except version & part),
+         guess: {
+             kind: movie | episode | trailer | interview | ...,
+             name: string,
+             year: number | null,
+             season?: number,
+             episodes?: number[],
+             absolutes?: number[],
+             externalId: Record<string, {showId, season, number}[]>,
+             remap: {
+                 from: "thexem",
+                 oldSeason: number,
+                 oldEpisodes: number[],
+              },
+             ...
+          },
+     }
+     ```
+ - If kind is episode, try to find the serie's id on kyoo (using the previously fetched data from `/videos`):
+   - if another video in the list of already registered videos has the same `kind`, `name` & `year`, assume it's the same
+   - if a match is found, add to the video's json:
+   ```json5
+   {
+       entries: (uuid | slug | {
+           show: uuid | slug,
+           season: number,
+           episode: number,
+           externalId?: Record<string, {showId, season, number}> // takes priority over season/episode for matching if we have one
+       })[],
+   }
+   ```
+ - Scanner pushes everything to the api in a single post `/videos` call
+ - Api registers every video in the database
+ - For each video without an associated entry, the guess data + the video's id is sent to the Matcher via a queue.
+ - Matcher retrieves metadata from the movie/serie + ALL episodes/seasons (from an external provider)
```

Comment:
How should we dedup metadata retrieval here? If we have 100 video of `One Piece`, we do not want to fetch every metadata of `One Piece` 100 times.
Should we group videos based on the guessed `name` & `year` fields before sending them as a bulk in the matcher's queue?

Reply:
The api would receive a bulk add of videos, and would then either:
- create an event per video
- or create an event per video group

This is not stateful, it only affects items sent in the bulk create request.

Adding a cache in the matcher would make things stateful & I feel might be more complex. With a cache we would need to:
- 1 check if item already exist on the api
  - if it does get it's id, link the video to it via an api call
- 2 check if an item of the same show is currently being processed:
  - if yes, wait for it to finish, retrieve it's id and go to step 1
 - 3 download metadata for item
 - 4 push metadata + link video with the metadata

vs just:
- download metadata for item
- push metadata + link videoS with the metadata

Reply:
I'm not sure I understand, wasn't the problem just getting the metadata multiple times?

So, for 100 videos:
- 1. Get metadata -> One Piece
- 2. Get metadata -> One Piece (cached)
and so on..